File: fizz-buzz/fizz_buzz_csapp_07/FizzBuzz.Tests/RendererTests.cs
namespace FizzBuzz.Tests;

public class RendererTests {
  int displayedCount = 0;
  readonly List<string> displayed = new();

  [Fact]
  public void ShowNumbers_DisplayAllNumbers() {
    Renderer renderer = new(Display);
    renderer.ShowNumbers(3, Render);
    Assert.Equal(3, displayedCount);
  }

  [Fact]
  public void ShowNumbers_RenderNumber() {
    Renderer renderer = new(Display);
    renderer.ShowNumbers(1, Render);
    Assert.Equal("_1", displayed[0]);
  }

  [Fact]
  public void FizzBuzz_FinalTest() {
    Renderer renderer = new(Display);
    renderer.ShowNumbers(100, FizzBuzz.Render);
    Assert.Equal("Fizz", displayed[2]);
    Assert.Equal("Buzz", displayed[4]);
    Assert.Equal("FizzBuzz", displayed[14]);
    Assert.Equal("16", displayed[15]);
    Assert.Equal("FizzBuzz", displayed[44]);
    Assert.Equal("Buzz", displayed[49]);
    Assert.Equal("Fizz", displayed[98]);
  }

  void Display(string text) {
    displayedCount++;
    displayed.Add(text);
  }

  string Render(int number) => $"_{number}";
}

File: fizz-buzz/fizz_buzz_pyapp_04/RendererTests.py
import unittest

from FizzBuzz import fizz_buzz
from Renderer import Renderer


class RendererTestCase(unittest.TestCase):
    def setUp(self):
        self.displayedCount = 0
        self.displayed = []
        self.renderer = Renderer(self.display)

    def display(self, text):
        self.displayedCount += 1
        self.displayed.append(text)

    def test_show_numbers_should_display_each_number(self):
        self.renderer.show_numbers(2, self.render)
        self.assertEqual(2, self.displayedCount)

    def test_show_numbers_should_render_number(self):
        self.renderer.show_numbers(1, self.render)
        self.assertEqual('f1', self.displayed[0])

    def test_fizz_buzz_final(self):
        self.renderer = Renderer(print)
        self.renderer.show_numbers(100, fizz_buzz)

    @staticmethod
    def render(number):
        return f'f{number}'


if __name__ == '__main__':
    unittest.main()
